addons/resource/models/resource.py

```diff
@@ -1020,6 +1025,54 @@ def _get_unavailable_intervals(self, start, end):
             resource_mapping.update(resources_unavailable_intervals)
         return resource_mapping
 
+    def _get_calendars_validity_within_period(self, start, end, default_company=None):
+        """
+            Returns a dict of dict with resource's id as first key and resource's calendar as secondary key
+            The value is the validity interval of the calendar for the given resource.
+
+            Here the validity interval for each calendar is the whole interval but it's meant to be overriden in further modules
+            handling resource's employee contracts.
+        """
+        assert start.tzinfo and end.tzinfo
+        calendars_within_period_per_resource = defaultdict(lambda: defaultdict(Intervals))  # keys are [resource id:integer][calendar:self.env['resource.calendar']]
+        default_calendar = default_company and default_company.resource_calendar_id or self.env.company.resource_calendar_id
+        if not self:
+            # if no resource, add the company resource calendar.
+            calendars_within_period_per_resource[False][default_calendar] = Intervals([(start, end, self.env['resource.calendar.attendance'])])
+        for resource in self:
+            calendar = resource.calendar_id or resource.company_id.resource_calendar_id or default_calendar
+            calendars_within_period_per_resource[resource.id][calendar] = Intervals([(start, end, self.env['resource.calendar.attendance'])])
+        return calendars_within_period_per_resource
+
+    def _get_work_intervals_batch(self, start, end, calendars=None):
```

Comment:
There is a confusion between ``_work_intervals_batch`` and ``_get_work_intervals_batch``.

I think you should find something more explicit for your method.

_work_intervals_batch returns all the presence intervals per resource.

Your method returns all the presence intervals:
- per resources (according to the calendar/contract validity)
- per calendar (which are the global presence intervals - resource_id = False)

Why not something like ``_get_valid_work_intervals_batch`` (not perfect I reckon, and the batch is not really needed, it's coming from the PR where we added new methods in stable 13.0 to improve the perf by making the computations in batch).

Reply:
I totally agree with it, `_get_valid_work_intervals` seems legit. If we had to suffix with `_batch` all the batched methods, we could live in a `_batch` world.